app/upgrades/v17/constants.go

```diff
@@ -17,3 +22,275 @@ var Upgrade = upgrades.Upgrade{
 		Deleted: []string{},
 	},
 }
+
+const (
+	QuoteAsset  = "uosmo"
+	TickSpacing = 100
+)
+
+type AssetPair struct {
+	BaseAsset         string
+	SpreadFactor      sdk.Dec
+	LinkedClassicPool uint64
+	Superfluid        bool
+}
+
+var AssetPairs = []AssetPair{
+	{LinkedClassicPool: 837},
+	{
+		SpreadFactor:      sdk.MustNewDecFromStr("0.0005"), // Normally 0.0002, but is not authorized
+		LinkedClassicPool: 857,
+	},
+	{LinkedClassicPool: 712},
+	{LinkedClassicPool: 773},
+	{LinkedClassicPool: 9},
+	{LinkedClassicPool: 3},
+	{LinkedClassicPool: 812},
+	{LinkedClassicPool: 584},
+	{LinkedClassicPool: 604},
+	{LinkedClassicPool: 497},
+	{LinkedClassicPool: 806},
+	{LinkedClassicPool: 907},
+	{LinkedClassicPool: 1013},
+	{LinkedClassicPool: 15},
+	{LinkedClassicPool: 586},
+	{LinkedClassicPool: 627},
+	{LinkedClassicPool: 795},
+	{LinkedClassicPool: 730},
+	{LinkedClassicPool: 7},
+	{LinkedClassicPool: 1039},
+	{LinkedClassicPool: 5},
+	{LinkedClassicPool: 573},
+	{LinkedClassicPool: 641},
+	{LinkedClassicPool: 605},
+	{LinkedClassicPool: 971},
+	{LinkedClassicPool: 625},
+}
+
+// AssetPairs contract: all AssetPairs being initialized in this upgrade handler all have the same quote asset (OSMO).
+func InitializeAssetPairs(ctx sdk.Context, keepers *keepers.AppKeepers) []AssetPair {
+	gammKeeper := keepers.GAMMKeeper
+	superfluidKeeper := keepers.SuperfluidKeeper
+	for i, assetPair := range AssetPairs {
+		pool, err := gammKeeper.GetCFMMPool(ctx, assetPair.LinkedClassicPool)
+		if err != nil {
+			panic(err)
+		}
+
+		// Set the base asset as the non-osmo asset in the pool
+		poolLiquidity := pool.GetTotalPoolLiquidity(ctx)
+		for _, coin := range poolLiquidity {
+			if coin.Denom != QuoteAsset {
+				AssetPairs[i].BaseAsset = coin.Denom
+				break
+			}
+		}
+
+		// If the spread factor is not manually set above, set it to the the same value as the pool's spread factor.
+		if assetPair.SpreadFactor.IsNil() {
+			AssetPairs[i].SpreadFactor = pool.GetSpreadFactor(ctx)
+		}
+
+		// Check if the pool is superfluid.
+		// If the pool is superfluid, set the superfluid flag to true.
+		poolShareDenom := fmt.Sprintf("gamm/pool/%d", assetPair.LinkedClassicPool)
+		_, err = superfluidKeeper.GetSuperfluidAsset(ctx, poolShareDenom)
+		if err == nil {
+			AssetPairs[i].Superfluid = true
+		}
```

Comment:
nit: what do you think of this approach? 
```suggestion
		if err != nil {
			continue
		}
		AssetPairs[i].Superfluid = true
```